lib/kino/debug.ex

```diff
@@ -0,0 +1,125 @@
+defmodule Kino.Debug do
+  @moduledoc false
+
+  @registry Kino.Debug.Registry
+
+  @doc """
+  Custom backend for `Kernel.dbg/2`.
+
+  The custom backend provides a more interactive user interface for
+  `Kernel.dbg/2` calls in certain cases, such as call pipelines. It
+  falls back to the default backend otherwise.
+  """
+  @spec dbg(Macro.t(), Macro.t(), Macro.Env.t()) :: Macro.t()
+  def dbg(ast, options, %Macro.Env{} = env) do
+    default_dbg_ast = Macro.dbg(ast, options, env)
+    dbg_id = System.unique_integer()
+
+    case ast do
+      {:|>, _meta, _args} ->
+        quote do
+          if pid = unquote(__MODULE__).lookup_dbg_handler(unquote(dbg_id)) do
+            send(pid, :dbg_call)
+            unquote(ast)
+          else
+            unquote(initialize_pipeline(ast, dbg_id, env))
+          end
+        end
+
+      _ ->
+        default_dbg_ast
+    end
+  end
+
+  defp initialize_pipeline(ast, dbg_id, env) do
+    [head_ast | rest_asts] = asts = for {ast, 0} <- Macro.unpipe(ast), do: ast
+
+    head_source = Macro.to_string(head_ast)
+    rest_sources = for ast <- rest_asts, do: "|> " <> Macro.to_string(ast)
+    sources = [head_source | rest_sources]
+
+    [head_var | rest_vars] = vars = Macro.generate_arguments(length(asts), __MODULE__)
+
+    assignments = quote do: unquote(head_var) = unquote(head_ast)
+
+    {assignments, _} =
+      rest_vars
+      |> Enum.zip(rest_asts)
+      |> Enum.reduce({assignments, head_var}, fn {var, node}, {assignments, prev_var} ->
+        assignments =
+          quote do
+            unquote(assignments)
+            unquote(var) = unquote(Macro.pipe(prev_var, node, 0))
+          end
+
+        {assignments, var}
+      end)
+
+    funs =
+      for ast <- rest_asts do
+        arg = Macro.var(:arg, __MODULE__)
+        expr = Macro.pipe(arg, ast, 0)
+
+        quote do
+          fn unquote(arg) -> unquote(expr) end
+        end
+      end
+
+    quote do
+      unquote(assignments)
```

Comment:
I don't think we need the assignments? It seems we only care about `head_var` and the result (the last var). In this case, I think we only need this:

```elixir
head_var = Macro.generate_unique_argument(:arg, __MODULE__)
result = Enum.reduce(rest_asts, head_var, &Macro.pipe(& &2, & &1, 0))

quote do
  unquote(head_var) = unquote(head_ast)
  unquote(__MODULE__).__dbg__(...)
  unquote(result)
end
```

Reply:
Furthermore, to avoid introducing the same AST multiple times on the dbg callback, instead of an if/else at the root, I think you actually want to move the `pid` handling here. So overall you would have this:

```
head_var = Macro.generate_unique_argument(:arg, __MODULE__)
result = Enum.reduce(rest_asts, head_var, &Macro.pipe(& &2, & &1, 0))

quote do
  unquote(head_var) = unquote(head_ast)

  unless lookup_pid(...) do
    wrapped_funs = fn -> unquote(funs) end
    unquote(__MODULE__).__dbg__(sources, head_var, ...)
  end

  unquote(result)
end
```

Reply:
@josevalim currently we pass all intermediate results to the kino, so for the first render we don't need to run all the functions. Do you think it's better to only pass the initial value and reevaluate?

Reply:
Hrm… this is tricky because the only times they will be different are when using random or when it depends on something from the current process.

it is probably best to show the original results then, so please ignore this particular bit. :) I would still move the pid lookup inside!